crates/kubelet/src/resources/device_plugin_manager/plugin_connection.rs

```diff
@@ -0,0 +1,320 @@
+use super::DeviceMap;
+use crate::device_plugin_api::v1beta1::{
+    device_plugin_client::DevicePluginClient, AllocateRequest, AllocateResponse, Device, Empty,
+    ListAndWatchResponse, RegisterRequest,
+};
+use std::collections::HashMap;
+use std::sync::Arc;
+use tokio::sync::{broadcast, RwLock};
+use tonic::Request;
+use tracing::{debug, error, trace};
+
+/// PluginConnection that maps to a single registered device plugin.
+/// It is responsible for managing gRPC communications with the device plugin and caching
+/// device states reported by the device plugin
+pub struct PluginConnection {
+    /// Client that is connected to the device plugin
+    client: DevicePluginClient<tonic::transport::Channel>,
+    /// `RegisterRequest` received when the device plugin registered with the DeviceRegistry
+    register_request: RegisterRequest,
+}
+
+impl PluginConnection {
+    pub fn new(
+        client: DevicePluginClient<tonic::transport::Channel>,
+        register_request: RegisterRequest,
+    ) -> Self {
+        PluginConnection {
+            client,
+            register_request,
+        }
+    }
+
+    pub async fn start(
+        &self,
+        devices: Arc<RwLock<DeviceMap>>,
+        update_node_status_sender: broadcast::Sender<()>,
+    ) {
+        match self
+            .list_and_watch(devices, update_node_status_sender)
+            .await
+        {
+            Err(e) => {
+                error!(error = %e, resource = %self.register_request.resource_name, "ListAndWatch ended unexpectedly for resource")
+            }
+            Ok(_) => {
+                trace!(resource = %self.register_request.resource_name, "Received message to stop ListAndWatch for resource")
+            }
+        }
+    }
+
+    /// Connects to a device plugin's ListAndWatch service.
+    /// The device plugin updates this client periodically about changes in the capacity and health of its resource.
+    /// Upon updates, this propagates any changes into the `plugins` map and triggers the `NodePatcher` to
+    /// patch the node with the latest values.
+    /// TODO: return error
+    async fn list_and_watch(
+        &self,
+        devices: Arc<RwLock<DeviceMap>>,
+        update_node_status_sender: broadcast::Sender<()>,
+    ) -> anyhow::Result<()> {
+        let mut stream = self
+            .client
+            .clone()
```

Comment:
the generated `list_and_watch`  function requires a mutable reference to the client, which cannot be derived from `&self`. The `PluginConnection` `list_and_watch` method cannot take in `&self` because the map of plugins is wrapped in an Arc.